packages/cli/src/metadataGeneration/metadataGenerator.ts

```diff
@@ -214,17 +214,34 @@ export class MetadataGenerator {
 
   public AddReferenceType(referenceType: Tsoa.ReferenceType) {
     if (!referenceType.refName) {
-      return;
+      throw new Error('no reference type name found');
     }
-    this.referenceTypeMap[decodeURIComponent(referenceType.refName)] = referenceType;
+    this.referenceTypeMap[referenceType.refName] = referenceType;
   }
 
   public GetReferenceType(refName: string) {
     return this.referenceTypeMap[refName];
   }
 
-  public OnFinish(callback: (referenceTypes: Tsoa.ReferenceTypeMap) => void) {
-    this.circularDependencyResolvers.push(callback);
+  public CheckModelUnicity(refName: string, positions: Array<{ fileName: string; pos: number }>) {
```

Comment:
Type unicity check:
* We want to check, that type A<B<...>,{ c: C}> is uniq or not. For this, we need to check if every A, B & C type identifier definitions are uniq or not. Uniq = A type can have more than one definition, but their definition positions array must be uniq. Position = fileName + pos. Not stored, so not interesting, which local machine the check is made.
* A<B> and A_B_ types are both renamed to A_B_ to use it in OpenApi. This is a second type of name conflicts.